src/notification/index.js

```diff
@@ -209,9 +209,15 @@ export class NotificationListener {
   };
 
   /** Private. */
-  handleDeviceToken = async (deviceToken: string) => {
-    this.dispatch(gotPushToken(deviceToken));
-    await this.dispatch(sendAllPushToken());
+  handleDeviceToken = async (deviceToken: string | { ... }) => {
+    // A device token should normally be a string of hex digits. Sometimes,
+    // however, we appear to receive objects here. Log this. (See issue #3672.)
+    if (typeof deviceToken === 'string' && deviceToken !== '[Object object]') {
+      this.dispatch(gotPushToken(deviceToken));
+      await this.dispatch(sendAllPushToken());
+    } else {
+      logging.error('Received non-string device token', { deviceToken });
+    }
```

Comment:
Let's write this error path in the early-return pattern:
```js
  if (isBad(thing)) {
    logging.error('Thing is bad', { thing });
    return;
  }

  processNormally(thing);
```

This keeps the code that handles each error next to the code that tests for it. A related effect is that it helps the reader to study the behavior in a given error case, or the behavior in the "normal"/happy case, or to understand the full behavior by divide-and-conquer.

Doesn't hugely matter for a function this short with just one error case. But it's good to have the pattern consistent for if the function grows, or if another contributor copies the pattern in another context. And even at this scale I think it already helps.

Reply:
(I actually did write it that way in the first draft; I switched to the above because the negated logic seemed more straightforward.)

Reply:
Cool. Yeah, I think the main thing that can make this nested style appealing is that it lets you write down the things that are true in the happy path, rather than the things that aren't.

But then I think the effect on the code's overall structure generally ends up outweighing that, even at this scale.

Thanks for the revision!